### DISCUSSION TITLE
System.Text.Json easier config for read/deserialize number as string

### DISCUSSION THREAD

**Discussion by hsenasilva** [dotnet/runtime#104281] [Ideas]
Hi, I was testing the System.Text.Json deserialization feature and I faced a scenario where double quoted number like "10.00" is parsed to System.Decimal [by default on ASP.NET Core apps](https://github.com/dotnet/docs/blob/main/docs/core/compatibility/serialization/5.0/jsonserializer-allows-reading-numbers-as-strings.md), and the opposite is not true. Which means that if we send an serialized decimal like 10.00 the default behavior is not the same, and the Converter will throw an exception. I think that could be useful an string implementation like [JsonNumberHandling](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/Common/JsonNumberHandling.cs#L13), where could be possible set an option on JsonSerializerOptions to permit parse automatically a number to string without the need to implement a custom converter. 

What are your thoughts about this idea? Any side effects on this more permissive approach? 

I could send an implementation proposal if you like the idea. o/

**Comment by huoyaoyuan** [comment]
> where could be possible set an option on JsonSerializerOptions to permit parse automatically a number to string

Isn't it covered by `JsonNumberHandling.WriteAsString`?

**Reply by mariussNG** [reply]
@huoyaoyuan It is not. This simple test throws an exception:

```c#
public class Test
{
    record class MyRecord(string Id);

    [Test]
    public void TestNumberToString()
    {
        const string json = """{"Id": 123}""";

        var jsonOptions = new JsonSerializerOptions { NumberHandling = JsonNumberHandling.WriteAsString };
        var deserialized = JsonSerializer.Deserialize<MyRecord>(json, jsonOptions);

        Console.WriteLine(deserialized);
    }
}
```

Exception:
```
System.Text.Json.JsonException : The JSON value could not be converted to Test+MyRecord. Path: $.Id | LineNumber: 0 | BytePositionInLine: 10.
  ----> System.InvalidOperationException : Cannot get the value of a token type 'Number' as a string.
```

Every other JSON-library I've used accepts this by default, so it is really surprising that System.Text.Json throws an exception, let alone have no simple option to allow it (without implementing a custom converter).

**Reply by huoyaoyuan** [reply]
> This simple test throws an exception:

This is the opposite case, reading JSON numbers as strings. It's not supported by System.Text.Json yet, as https://learn.microsoft.com/dotnet/standard/serialization/system-text-json/migrate-from-newtonsoft?pivots=dotnet-9-0#non-string-values-for-string-properties .

**Comment by AsifZaman777** [comment]
if "10.00" (as a string) can deserialize to a decimal, we should be able to serialize 10.00m back to "10.00" without a custom converter.
Something like this would be super handy:
```
var options = new JsonSerializerOptions
{
    NumberHandling = JsonNumberHandling.WriteNumbersAsStrings
};

var json = JsonSerializer.Serialize(10.00m, options);
// Outputs: "10.00"
```
As long as it’s opt-in, I don’t see any major issues. It would make interop with other systems much easier. I’d support a proposal for this!

**Comment by dfaivre-pcs** [comment]
I'd love to see a fix for this!
